openmmtools/forcefactories.py

```diff
@@ -110,22 +121,41 @@ def restrain_atoms(thermodynamic_state, sampler_state, topology,
     K = thermodynamic_state.kT / sigma**2  # Spring constant.
     system = thermodynamic_state.system  # This is a copy.
 
-    # Make sure the topology is an MDTraj topology.
-    if isinstance(topology, mdtraj.Topology):
-        mdtraj_topology = topology
-    else:
-        mdtraj_topology = mdtraj.Topology.from_openmm(topology)
-
-    # Translate the system to the origin to avoid
-    # MonteCarloBarostat rejections (see openmm#1854).
-    protein_atoms = mdtraj_topology.select('protein')
-    distance_unit = sampler_state.positions.unit
-    centroid = np.mean(sampler_state.positions[protein_atoms,:] / distance_unit, 0) * distance_unit
-    sampler_state.positions -= centroid
+    # Check that there are atoms to restrain.
+    if len(restrained_atoms) == 0:
+        raise ValueError('No atoms to restrain.')
+
+    # We need to translate the restrained molecule to the origin
+    # to avoid MonteCarloBarostat rejections (see openmm#1854).
+    if thermodynamic_state.pressure is not None:
+        # First, determine all the molecule atoms. Reference platform is the cheapest to allocate?
+        reference_platform = openmm.Platform.getPlatformByName('Reference')
+        integrator = openmm.VerletIntegrator(1.0*unit.femtosecond)
+        context = openmm.Context(system, integrator, reference_platform)
+        molecules_atoms = context.getMolecules()
```

Comment:
Do we have any custom/alchemical forces relying on a `CustomCompoundBondForce` or a `CustomBond` force to replicate a force which is NOT a molecule? e.g. if we have multiple restraints such as in the future, this could cause problems on the second pass of restraints